### PR TITLE
[MM-20520] Migrate 'components/new_channel_flow' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by kosgrz**
#### Summary
This PR migrates 'components/new_channel_flow' module to TypeScript.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/13720

### CODE REVIEW

File: components/new_channel_flow/new_channel_flow.tsx [hunk 1]
```diff
@@ -116,17 +141,27 @@ export default class NewChannelFlow extends React.Component {
...
+            group_constrained: false,
+            id: '',
+            last_post_at: 0,
+            scheme_id: '',
+            total_msg_count: 0,
+            update_at: 0,
         };
 
-        actions.createChannel(channel).then((result) => {
+        actions.createChannel(channel, currentUserId).then((result: {data: any; error?: any}) => {
```
Comment (reflog) [inline components/new_channel_flow/new_channel_flow.tsx]: Can we avoid 'any' here? Maybe there's appropriate type returned from createChannel in redux?

File: components/new_channel_flow/new_channel_flow.tsx [hunk 2]
```diff
@@ -137,7 +172,7 @@ export default class NewChannelFlow extends React.Component {
         });
     };
 
-    onCreateChannelError = (err) => {
+    onCreateChannelError = (err: {id: string; message: string}) => {
```
Comment (reflog) [inline components/new_channel_flow/new_channel_flow.tsx]: I think we have an error type in mattermost-redux...

File: components/new_channel_flow/new_channel_flow.tsx [hunk 3]
```diff
@@ -192,7 +227,7 @@ export default class NewChannelFlow extends React.Component {
         }
     };
 
-    render() {
+    render = (): JSX.Element => {
```
Comment (reflog) [inline components/new_channel_flow/new_channel_flow.tsx]: why the binding here?

File: components/new_channel_flow/new_channel_flow.tsx [hunk 4]
```diff
@@ -1,78 +1,103 @@
...
+     */
+    canCreatePublicChannel: boolean;
+
+    /**
+     * Permission to create private channel
+     */
+    canCreatePrivateChannel: boolean;
+
+    actions: {
+        createChannel: (channel: Channel, userId: string) => Promise<{data: {id: string; name: string}; error?: any}>;
```
Comment (reflog) [inline components/new_channel_flow/new_channel_flow.tsx]: we should avoid `any`

File: components/new_channel_flow/index.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import NewChannelFlow from './new_channel_flow';
+
+type Actions = {
+    createChannel: (channel: Channel, userId: string) => Promise<{data: {id: string; name: string}; error?: any}>;
+    switchToChannel: (channel: any) => Promise<{}>;
+}
+
+function mapStateToProps(state: GlobalState) {
+    return {
+        currentUserId: getCurrentUserId(state),
```
Comment (reflog) [inline components/new_channel_flow/index.ts]: why was this added? how did it work before the TS conversion?

File: components/new_channel_flow/new_channel_flow.tsx (hunk 1 above)
Comment (kosgrz) [reply components/new_channel_flow/new_channel_flow.tsx]: Done :+1:

File: components/new_channel_flow/new_channel_flow.tsx (hunk 2 above)
Comment (kosgrz) [reply components/new_channel_flow/new_channel_flow.tsx]: Done :+1:

File: components/new_channel_flow/new_channel_flow.tsx (hunk 3 above)
Comment (kosgrz) [reply components/new_channel_flow/new_channel_flow.tsx]: Sorry, by a mistake. I've just switched it to a normal function.

File: components/new_channel_flow/new_channel_flow.tsx (hunk 4 above)
Comment (kosgrz) [reply components/new_channel_flow/new_channel_flow.tsx]: Done :+1:

File: components/new_channel_flow/index.ts (hunk 5 above)
Comment (kosgrz) [reply components/new_channel_flow/index.ts]: I thought it's required by the "createChannel" action, but it works fine without this param :+1:
